src/inc/msquic.hpp

```diff
@@ -805,6 +803,15 @@ struct MsQuicConnection {
         MsQuic->ConnectionShutdown(Handle, Flags, ErrorCode);
     }
 
+    void
+    Close(
+    ) noexcept {
+        if (Handle) {
+            MsQuic->ConnectionClose(Handle);
+            Handle = nullptr;
+        }
```

Comment:
This function needs synchronization. It could be called from either the MsQuic callback thread or from the test thread, in parallel. Linux tests show this crash:
```
##[warning][06/29/2022 08:10:28] Core file(s) generated
50	../sysdeps/unix/sysv/linux/raise.c: No such file or directory.
==================================================================================
 msquictest.9642.1656490227.core
==================================================================================
[Current thread is 1 (Thread 0x7f25ea2f0700 (LWP 9649))]
#0  __GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:50
#1  0x00007f25f39b0859 in __GI_abort () at abort.c:79
#2  0x00007f25f641a2e2 in __sanitizer::Abort () at ../../../../src/libsanitizer/sanitizer_common/sanitizer_posix_libcdep.cc:155
#3  0x00007f25f6424e8c in __sanitizer::Die () at ../../../../src/libsanitizer/sanitizer_common/sanitizer_termination.cc:57
#4  0x00007f25f640652c in __asan::ScopedInErrorReport::~ScopedInErrorReport (this=0x7f25ea2eef16, __in_chrg=<optimized out>) at ../../../../src/libsanitizer/asan/asan_report.cc:185
#5  0x00007f25f6405fa3 in __asan::ReportGenericError (pc=94110799472651, bp=bp@entry=139800819465104, sp=sp@entry=139800819465088, addr=addr@entry=106721347436352, is_write=is_write@entry=true, access_size=access_size@entry=8, exp=0, fatal=true) at ../../../../src/libsanitizer/asan/asan_report.cc:458
#6  0x00007f25f640738e in __asan::__asan_report_store8 (addr=addr@entry=106721347436352) at ../../../../src/libsanitizer/asan/asan_rtl.cc:124
#7  0x00005597e0c5a80b in MsQuicConnection::Close (this=0x61100000ff40) at /home/vsts/work/1/s/src/inc/msquic.hpp:811
#8  CloseFromCallbackContext::Callback (Conn=0x61100000ff40, Context=<optimized out>, Event=<optimized out>) at /home/vsts/work/1/s/src/test/lib/ApiTest.cpp:1172
#9  0x00005597e0c5d0aa in MsQuicConnection::MsQuicCallback (pThis=0x61100000ff40, Event=0x7f25ea2efca0) at /home/vsts/work/1/s/src/inc/msquic.hpp:1095
#10 0x00007f25f51b942d in QuicConnIndicateEvent (Connection=Connection@entry=0x61f00000fc80, Event=Event@entry=0x7f25ea2efca0) at /home/vsts/work/1/s/src/core/connection.c:730
#11 0x00007f25f518c93e in QuicWorkerProcessConnection (Worker=Worker@entry=0x6200000007c0, Connection=0x61f00000fc80, ThreadID=ThreadID@entry=9649, TimeNow=TimeNow@entry=0x7f25ea2efde0) at /home/vsts/work/1/s/src/core/worker.c:503
#12 0x00007f25f518eba4 in QuicWorkerLoop (Context=Context@entry=0x6200000007c0, TimeNow=TimeNow@entry=0x7f25ea2efde0, ThreadID=ThreadID@entry=9649) at /home/vsts/work/1/s/src/core/worker.c:670
#13 0x00007f25f518f5ff in QuicWorkerThread (Context=0x6200000007c0) at /home/vsts/work/1/s/src/core/worker.c:735
#14 0x00007f25f3b88609 in start_thread (arg=<optimized out>) at pthread_create.c:477
#15 0x00007f25f3aad133 in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:95
##[error][06/29/2022 08:10:30] Basic.ConnectionCloseFromCallback failed:
```